migration/migrator.go

```diff
@@ -0,0 +1,353 @@
+		mu.Lock()
+		defer mu.Unlock()
+		if err := mpt.UpdateStorage(address, slot, trimmed); err != nil {
+			return err
+		}
+		log.Trace("Updated storage slot to MPT", "contract", address.Hex(), "index", common.BytesToHash(key).Hex())
+		return nil
+	})
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	root, err := m.commit(mpt, types.EmptyRootHash)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	log.Debug("Storage migration finished", "account", address, "slots", slots.Load(), "elapsed", time.Since(startAt))
+	return root, nil
+}
+
+func (m *StateMigrator) readZkPreimage(key []byte) []byte {
+	hk := *trie.IteratorKeyToHash(key, true)
+	if preimage, ok := m.allocPreimage[hk]; ok {
+		return preimage
+	}
+	if preimage := m.zkdb.Preimage(hk); preimage != nil {
+		if common.BytesToHash(zk.MustNewSecureHash(preimage).Bytes()).Hex() == hk.Hex() {
+			return preimage
+		}
+	}
+	panic("preimage does not exist: " + hk.Hex())
+}
+
+func (m *StateMigrator) commit(mpt *trie.StateTrie, parentHash common.Hash) (common.Hash, error) {
+	root, set, err := mpt.Commit(true)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if set == nil {
+		log.Warn("Tried to commit state changes, but nothing has changed.", "root", root)
+		return root, nil
+	}
+
+	var hashCollidedErr error
+	set.ForEachWithOrder(func(path string, n *trienode.Node) {
+		if hashCollidedErr != nil {
+			return
+		}
+		// NOTE(pangssu): It is possible that the keccak256 and poseidon hashes collide, and data loss can occur.
+		data, _ := m.db.Get(n.Hash.Bytes())
+		if len(data) == 0 {
+			return
+		}
+		if node, err := zk.NewTreeNodeFromBlob(data); err == nil {
+			hashCollidedErr = fmt.Errorf("Hash collision detected: hashKey: %v, key: %v, value: %v, zkNode: %v", n.Hash.Bytes(), path, data, node)
+		}
+	})
+	if hashCollidedErr != nil {
+		return common.Hash{}, hashCollidedErr
+	}
+
+	if err := m.mptdb.Update(root, parentHash, 0, trienode.NewWithNodeSet(set), nil); err != nil {
+		return common.Hash{}, err
+	}
+	if err := m.mptdb.Commit(root, false); err != nil {
+		return common.Hash{}, err
+	}
+	return root, nil
+}
+
+func (m *StateMigrator) FinalizeTransition(transitionBlock types.Block) {
+	// We need to update the chain config to set the correct hardforks.
+	genesisHash := rawdb.ReadCanonicalHash(m.db, 0)
+	cfg := rawdb.ReadChainConfig(m.db, genesisHash)
+	if cfg == nil {
+		panic("chain config not found")
+	}
+
+	// Set the standard options.
+	cfg.LondonBlock = transitionBlock.Number()
+	cfg.ArrowGlacierBlock = transitionBlock.Number()
+	cfg.GrayGlacierBlock = transitionBlock.Number()
+	cfg.MergeNetsplitBlock = transitionBlock.Number()
+	cfg.TerminalTotalDifficulty = big.NewInt(0)
+	cfg.TerminalTotalDifficultyPassed = true
+
+	// Set the Optimism options.
+	cfg.BedrockBlock = transitionBlock.Number()
+	// Enable Regolith from the start of Bedrock
+	cfg.RegolithTime = new(uint64)
+	// Switch KromaConfig to OptimismConfig
+	cfg.Optimism = &params.OptimismConfig{
```

Comment:
I'm not sure exactly what your intentions were in adding OptimismConfig, is it a step to reduce code changes?